Improve styler

jecisc opened the pull request:
This PR contains multiple changes:
- Introduction of an infrastructure to better test the styler that do not rely on the colors of the current theme
- Add tests on multiple elements to style (I'll add more soon)
- Fix bug where argument was detected as a temporary variable
- Fix bug where block pattern temp var was detected as a temporary variable
- Fix bug where shared pools vars were detected as class variables
- Fix bug where class var is considered a global for style
- Remove multiple style that are not used by Shout such as: excessCode, unfinishedComment, unfinishedString, binary, unary, $, -, stringSymbol, literalArray, methodArg, workspaceVar, incompleteBinary, incompleteUnary, undefinedBinary, undefinedUnary, arrayStart, arrayEnd, arrayStart1, arrayEnd1, externalCallType, externalCallTypePointerIndicator, primitiveOrExternalCallStart, primitiveOrExternalCallEnd, primitive, pragmaUnary, pragmaBinary, externalFunctionCallingConvention, module

Fixes #12925 
Fixes #12914

jecisc committed:
Add test infrastructure to styler

jecisc committed:
Add nore tests on shout styling

jecisc committed:
Remove excessCode style that is never used 

Shout never assigns any style of that name

jecisc committed:
Add tests on comments styling

jecisc committed:
Remove unused style unfinishedComment

It is never assigned to any node by Shout

jecisc committed:
Test character style + remove unused '$'style

jecisc committed:
Test number style + remove unused integer style

jecisc committed:
Remove unused style #-

jecisc committed:
Remove StringSymbol style that is never used

jecisc committed:
Remove literal array style that is unused

jecisc committed:
Remove unused unfinishedString style

jecisc committed:
Remove some assignment style that are unused

jecisc committed:
Remove unused literal style

jecisc committed:
Remove unused styles #unari and #binary

We should rename #keyword into #message later

jecisc committed:
Add some shout tests

jecisc committed:
Fix bug where argument style was not set correctly

jecisc committed:
Remove #methodArg style since it seems unreacheable

It was probably replaced by #patternArg

jecisc committed:
Fixes some missmatch for arguments

jecisc committed:
Fixe bug in block pattern temp var style

jecisc committed:
Remove unused workspace var style

jecisc committed:
Fix bug where shared pools don't have the right style

jecisc committed:
Fix bug where class var is considered a global for style

jecisc committed:
More style tests + remove unused incompleteUnary and incompleteBinary

jecisc committed:
Add tests + remove unused undefinedBinary and undefinedUnary

MarcusDenker commented:
- SpCodeAdapterTest fails, it seems to nedd #unary 
- NoUnusedVariablesLeftTest>>testNoUnusedClassVariablesLeft

jecisc committed:
Add test + remove unused styles #patternUnary and #patternBinary

MarcusDenker commented:
Full image load failure, but that might not be due to this PR?

```
MetacelloNotification: Loading -> Chest-tonel.1 --- github://pharo-spec/Chest:v0.4.2 --- cache
Error: Package Chest depends on the following classes:
  StPlaygroundInteractionModel
  StDebugger
  StDebuggerExtensionVisitor
  StDebuggerExtensionInspectorNodeBuilderVisitor
You must resolve these dependencies before you will be able to load these definitions: 
  ChestCodeScriptingInteractionModel
  ChestCodeScriptingInteractionModel>>#debugger:
  ChestCodeScriptingInteractionModel>>#addBinding:
  ChestCodeScriptingInteractionModel>>#addSelfAndThisContextBindings
  ChestCodeScriptingInteractionModel>>#bindingOf:
  ChestCodeScriptingInteractionModel>>#bindings
  ChestCodeScriptingInteractionModel>>#debugger:
  ChestCodeScriptingInteractionModel>>#hasBindingOf:
  ChestCodeScriptingInteractionModel>>#removeBinding:
  StDebugger>>#bindings
  StDebugger>>#interactionModel
  StDebuggerExtensionInspectorNodeBuilderVisitor>>#visitChest:
  StDebuggerExtensionVisitor>>#visitChest:
```

jecisc committed:
Add tests + remove array related styles that are not used anymore

jecisc committed:
Add tests

privat commented:
Nice!

> The goal is to test more things, remove all unused styles, maybe rename #keyword into #selector and delegate more of the style application to the AST

You had my curiosity, but now you have my attention. What are your plan about the AST?

jecisc committed:
Add tests about pragma + Remove unused styles:
externalCallType, externalCallTypePointerIndicator, primitiveOrExternalCallStart, primitiveOrExternalCallEnd, primitive, pragmaUnary, pragmaBinary, externalFunctionCallingConvention, module

jecisc commented:
I just finished to:
- Remove all unused style
- Fix bug where wrong style was assigned (we did not see it because the different style share the same color **in the current theme**)
- Add a test for **every** style that is assigned by Shout currently so that we do not have regressions
I updated my first comment in order to list all the changes I did

jecisc commented:
I still need to fix the failing tests

privat commented:
Also, It might interest you, but I added more than 250 RBCodeSnippet to systematically test various buggy pieces of code ­— and legitimate non-buggy pieces of code, but those are boring :)

I added a simple test to run the styler on all those pieces of code. Basically it just checks that no internal exception like MNU or other thing are fired when dealing with faulty AST. https://github.com/pharo-project/pharo/blob/Pharo11/src/Shout-Tests/RBCodeSnippetTest.extension.st and https://github.com/pharo-project/pharo/blob/Pharo11/src/Shout-Tests/RBCodeSnippet.extension.st

Do you have any idea to test more than "execution went fine" and check that the result is consistent? Annotating 250 code snippets with an expected styler result seems to be a menial task. Maybe there is something that is easier to check automatically.

jecisc commented:
In this PR I added some utilites to help test the style of a method.

Before, shout was testing the color of the styled nodes. 

In this PR, I added a Mock styler that will not change the style (the colors, emphasis, ...), but it tags the characters with the style that was found for them (instVar, classVar, pragma, ...).

If you want to test really the style that is applied to some elements of your snippets you can check SHRBStyleAttributionTest that I added to this PR

jecisc committed:
Fix release test

jecisc committed:
Add comments

jecisc commented:
testGmailEncrypted

demarey commented on src/OSWindow-SDL2/SDL_TextEditingEvent.class.st:
```diff
@@ -138,24 +138,22 @@ SDL_TextEditingEvent >> start: anObject [
-		[ :i |
-		(self getHandle nbUInt8AtOffset: i) = 0
-			ifTrue: [ ^ len ].
+	12 to: 12 + 32 - 1 do: [ :i |
```
Argh. All these magic numbers are really bad!
(I know it is not related to this PR)

jecisc replied:
I avoided to commit them multiple time but I forgot once :( should be pretty safe tho

demarey commented on src/Shout-Tests/SHRBMockSharedPoolStyler.class.st:
```diff
@@ -0,0 +1,11 @@
+I am a shared pool used for styler tests.
+"
+Class {
+	#name : #SHRBMockSharedPoolStyler,
```
I do not see where it is used

jecisc replied:
It is used as a shared pool in the test class to be able to test shared pool style

jecisc replied:
![0FC59509-323C-4181-A181-FB79A0D451D6](https://user-images.githubusercontent.com/9519971/223670113-674d10fd-52b5-46ae-a4a1-79aa3587b18e.jpeg)

demarey commented on src/Shout-Tests/SHRBTextStylerTest.class.st:
```diff
@@ -31,96 +31,13 @@ SHRBTextStylerTest >> style: aText [
 	super tearDown
 ]
 
-{ #category : #tests }
```
Where are the new tests equivalent to those removed?

jecisc replied:
Yes, but not relying on the current theme! Some of those were in reality failing but we didn’t see it since it was relying on the theme

demarey replied:
ah, sorry, I did not see the `load diff` hiding a lot of changes.
